src/3d/qgsraycastingutils_p.h

```diff
@@ -104,6 +107,40 @@ namespace QgsRayCastingUtils
                                 QVector3D c,
                                 QVector3D &uvw,
                                 float &t );
+
+  /**
+   * Helper struct to store ray casting results.
+   */
+  struct RayHit
+  {
+    //! Constructor
+    RayHit( const float distance, const QVector3D pos, const QgsFeatureId fid = FID_NULL, const QVariantMap attributes = QVariantMap() )
+      : distance( distance )
+      , pos( pos )
+      , fid( fid )
+      , attributes( attributes )
+    {
+    }
+    float distance;  //!< Distance from ray's origin
+    QVector3D pos;  //!< Hit position in world coordinates
+    QgsFeatureId fid;  //!< Fid of feature hit closest to ray origin, FID_NULL if no features hit
+    QVariantMap attributes;  //!< Point cloud point attributes, empty map if no point cloud points hit
+  };
+
+  /**
+   * Helper struct to store ray casting parameters.
+   */
+  struct RayCastContext
+  {
+    RayCastContext( bool singleResult = true, QSize screenSize = QSize(), float maxDistance = 1 )
+      : singleResult( singleResult )
+      , screenSize( screenSize )
+      , maxDistance( maxDistance )
+    {}
+    bool singleResult;  //!< If set to TRUE, only the closest point cloud hit will be returned (other entities always return only closest hit)
+    QSize screenSize;  //!< QSize of the 3d engine window
+    float maxDistance;  //!< The maximum distance from ray origin to look for hits when casting a ray
```

Comment:
worth adding an extra note to maxDistance that it would be normally set to far plane (to ignore data that would not get displayed in the 3D view)

Reply:
fixed